tests/run/pure_py.py

```diff
@@ -319,6 +319,32 @@ def test_cdef_nogil(x):
     return result
 
 
+@cython.cfunc
+@cython.inline
+def has_inner_func1(x):
+    # the inner function must remain a Python function
+    # (and inline must not be applied to it)
+    def inner():
+        return x
+    return inner
+
+
+with cython.cfunc:
+    def has_inner_func2(x):
+        # cfunc is not applied to "inner"
+        def inner():
+            return x
+        return inner
```

Comment:
Not sure if this is enough. It probably is, currently, since we do not allow `cdef` functions as inner functions. But that may not be the case forever, and there's the auto-wrapping feature which might strike then. I think a tree assertion would make this safer.

Reply:
The tree assertion looks to break it (since it creates a CompilerDirectivesNode, which has the old directives still attached).

I think this is why I found this tricky last time I tried to look at this - `AdjustDefByDirectives` is too late because directives are already set up everywhere, but it's the logical place to do it since it's where you know that the `cclass/cfunc` compiler directives are being applied